File: figures/08-graphs/snippet-topo-kahn.java
    public static Optional<int[]> kahn(Graph g) {
        int n = g.vertices(), used = 0;
        int[] indegree = new int[n], out = new int[n];
        for (long[] e : g.arcs()) indegree[(int)e[1]]++;
        PriorityQueue<Integer> zero = new PriorityQueue<>();
        for (int v = 0; v < n; v++) if (indegree[v] == 0) zero.add(v);
        while (!zero.isEmpty()) {
            int u = zero.remove(); out[used++] = u;
            for (Graph.Edge e : g.edgesFrom(u)) if (--indegree[e.to()] == 0) zero.add(e.to());
        }
        return used == n ? Optional.of(out) : Optional.empty();
    }

File: figures/08-graphs/snippet-topo-kahn.py
def kahn(g):
    n = g.vertices()
    indegree = [0] * n
    for _, v, _ in g.arcs():
        indegree[v] += 1
    zero = [v for v in range(n) if indegree[v] == 0]
    heapq.heapify(zero)
    order = []
    while zero:
        u = heapq.heappop(zero)
        order.append(u)
        for v, _ in g.edges_from(u):
            indegree[v] -= 1
            if indegree[v] == 0:
                heapq.heappush(zero, v)
    return order if len(order) == n else None
